testing/test_deploy.py

```diff
@@ -0,0 +1,218 @@
+  namespace.api_version = "v1"
+  namespace.kind = "Namespace"
+  namespace.metadata = k8s_client.V1ObjectMeta(name=run_label, labels={
+    "app": "kubeflow-e2e-test",
+    }
+  )
+
+  try:
+    logging.info("Creating namespace %s", namespace.metadata.name)
+    api.create_namespace(namespace)
+  except rest.ApiException as e:
+    if e.status == 409:
+      logging.info("Namespace %s already exists.", namespace.metadata.name)
+    else:
+      raise
+
+  return namespace
+
+def setup(args):
+  """Test deploying Kubeflow."""
+  project = args.project
+  cluster_name = args.cluster
+  zone = args.zone
+  util.configure_kubectl(project, zone, cluster_name)
+
+  util.load_kube_config()
+  # Create an API client object to talk to the K8s master.
+  api_client = k8s_client.ApiClient()
+
+  now = datetime.datetime.now()
+  run_label = "e2e-" + now.strftime("%m%d-%H%M-") + uuid.uuid4().hex[0:4]
+
+  if not os.path.exists(args.test_dir):
+    os.makedirs(args.test_dir)
+
+  logging.info("Using test directory: %s", args.test_dir)
+
+  namespace_name = run_label
+  def run():
+    namespace = _setup_test(api_client, namespace_name)
+    logging.info("Using namespace: %s", namespace)
+    # Set a GITHUB_TOKEN so that we don't rate limited by GitHub;
+    # see: https://github.com/ksonnet/ksonnet/issues/233
+    os.environ["GITHUB_TOKEN"] = args.github_token
+
+    # Initialize a ksonnet app.
+    app_name = "kubeflow-test"
+    util.run(["ks", "init", app_name,], cwd=args.test_dir)
+
+    app_dir = os.path.join(args.test_dir, app_name)
+
+    # TODO(jlewi): In presubmits we probably want to change this so we can
+    # pull the changes on a branch. Its not clear whether that's well supported
+    # in Ksonnet yet.
+    kubeflow_registry = "github.com/google/kubeflow/tree/master/kubeflow"
+    util.run(["ks", "registry", "add", "kubeflow", kubeflow_registry], cwd=app_dir)
+
+    # Install required packages
+    # TODO(jlewi): For presubmits how do we pull the package from the desired
+    # branch at the desired commit.
+    packages = ["kubeflow/core", "kubeflow/tf-serving", "kubeflow/tf-job"]
+
+    for p in packages:
+      util.run(["ks", "pkg", "install", p], cwd=app_dir)
+
+    # Deploy Kubeflow
+    util.run(["ks", "generate", "core", "kubeflow-core", "--name=kubeflow-core",
+              "--namespace=" + namespace.metadata.name], cwd=app_dir)
+
+    util.run(["ks", "apply", "default", "-c", "kubeflow-core",], cwd=app_dir)
+
+    # Verify that the TfJob operator is actually deployed.
+    tf_job_deployment_name = "tf-job-operator"
+    logging.info("Verifying TfJob controller started.")
+    util.wait_for_deployment(api_client, namespace.metadata.name,
+                             tf_job_deployment_name)
+
+    # Verify that JupyterHub is actually deployed.
+    jupyter_name = "tf-hub"
+    logging.info("Verifying TfHub started.")
+    util.wait_for_statefulset(api_client, namespace.metadata.name, jupyter_name)
```

Comment:
Do we want this to create the loadbalancer service as well (I think that was the old default behavior)? If that's still the case, it might be better to turn that off, or create a ClusterIP type of service.

Reply:
I think the correct thing for the test to do is to deploy Kubeflow with the default options. I think the question about what are the right defaults for Jupyter is valid but orthogonal. I opened #80.